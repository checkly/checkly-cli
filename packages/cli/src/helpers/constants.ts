export const DEFAULT_REGION = 'eu-central-1'
