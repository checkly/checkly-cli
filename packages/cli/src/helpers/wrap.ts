export interface WrapOptions {
  length?: number
  prefix?: string
  trimPrefix?: boolean
}

export function wrap (
  content: string,
  { length = 80, prefix = '', trimPrefix = true }: WrapOptions = {},
): string {
  const tokens = content.split(/([ \t\r\n])/)
  const lines = []
  let line = ''
  let trim = 0

  const prefixTrim = trimPrefix
    ? (prefix.match(/[ \t]$/) ?? [''])[0].length
    : 0

  for (const token of tokens) {
    if (token === '') {
      // Between tokens, not useful, skip.
      continue
    }

    if (token === '\r') {
      // Remove entirely.
      continue
    }

    if (token === ' ' || token === '\t') {
      // Add indiscriminately. These will get trimmed if needed.
      line += token
      trim += 1
      continue
    }

    if (token === '\n') {
      // Respect line endings.
      lines.push(line.slice(0, line.length - trim))
      line = prefix
      trim = prefixTrim
      continue
    }

    if (line.length === 0) {
      // We can assume that the first token of a line will always fit. Even
      // if it does not, it would not fit on any other line either.
      line += prefix
      line += token
      continue
    }

    if (line.length + token.length > length) {
      // Line would get too long, start a new line.
      lines.push(line.slice(0, line.length - trim))
      line = prefix
    }

    trim = 0
    line += token
  }

  lines.push(line.slice(0, line.length - trim))

  return lines.join('\n')
}
