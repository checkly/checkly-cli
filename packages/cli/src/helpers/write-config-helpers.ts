import * as recast from 'recast'
import path from 'node:path'
import fs from 'node:fs/promises'
import { namedTypes as n } from 'ast-types';


export function findPropertyByName (ast: any, name: string): recast.types.namedTypes.Property | undefined {
  let node
  recast.visit(ast, {
    visitProperty (path: any) {
      if (path.node.key.name === name) {
        node = path.node
      }
      return false
    },
  })
  return node
}

export function addOrReplaceItem(ast: any, node: any, name: string) {
  const item = findPropertyByName(ast, name)
  if (item) {
    item.value = node.value
  } else {
    ast.properties.push(node)
  }
}

export function addItemToArray(ast: any, node: any, name: string) {
  if (!n.ObjectExpression.check(ast)) {
    throw new Error('AST node is not an ObjectExpression');
  }

  const item = findPropertyByName(ast, name);

  if (item) {
    if (n.ArrayExpression.check(item.value)) {
      item.value.elements = item.value.elements || [];
      item.value.elements.push(node);
    } else {
      throw new Error(`Property "${name}" exists but is not an array`);
    }
  } else {
    ast.properties.push({
      type: 'Property',
      key: { type: 'Identifier', name },
      value: { type: 'ArrayExpression', elements: [node] },
      kind: 'init',
      computed: false,
      method: false,
      shorthand: false,
    });
  }
}


export async function reWriteChecklyConfigFile (data: string, fileName: string, dir: string) {
  await fs.writeFile(path.join(dir, fileName), data)
}
