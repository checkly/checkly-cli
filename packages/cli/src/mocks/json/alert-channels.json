[
  {
    "id": 240783,
    "type": "SMS",
    "config": {
      "number": "+35612345678"
    },
    "created_at": "2024-12-02T10:18:24.012Z",
    "updated_at": "2024-12-05T10:54:20.983Z",
    "sendRecovery": false,
    "sendFailure": true,
    "sendDegraded": false,
    "sslExpiry": true,
    "sslExpiryThreshold": 25,
    "autoSubscribe": false,
    "subscriptions": [
      {
        "id": 12799454,
        "checkId": "dde39bfb-f85a-4546-84d7-4085623dc4c9",
        "activated": true,
        "groupId": null
      }
    ]
  },
  {
    "id": 240784,
    "type": "EMAIL",
    "config": {
      "address": "example@gmail.com"
    },
    "created_at": "2024-12-02T10:18:24.020Z",
    "updated_at": "2024-12-05T10:54:20.988Z",
    "sendRecovery": false,
    "sendFailure": true,
    "sendDegraded": false,
    "sslExpiry": true,
    "sslExpiryThreshold": 25,
    "autoSubscribe": false,
    "subscriptions": [
      {
        "id": 12799455,
        "checkId": "dde39bfb-f85a-4546-84d7-4085623dc4c9",
        "activated": true,
        "groupId": null
      }
    ]
  },
  {
    "id": 240593,
    "type": "EMAIL",
    "config": {
      "address": "example@gmail.com"
    },
    "created_at": "2024-11-28T17:24:33.252Z",
    "updated_at": null,
    "sendRecovery": true,
    "sendFailure": true,
    "sendDegraded": false,
    "sslExpiry": false,
    "sslExpiryThreshold": 30,
    "autoSubscribe": false,
    "subscriptions": []
  }
]
