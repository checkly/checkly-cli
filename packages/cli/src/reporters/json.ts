import * as fs from 'fs'
import * as path from 'path'

import AbstractListReporter, { checkFilesMap } from './abstract-list'
import { CheckRunId, SequenceId } from '../services/abstract-check-runner'
import { printLn, getTestSessionUrl } from './util'

const outputFile = './checkly-json-report.json'

type JsonBuilderOptions = {
  testSessionId?: string
  numChecks: number
  runLocation: string
  checkFilesMap: checkFilesMap
}

export class JsonBuilder {
  testSessionId?: string
  numChecks: number
  runLocation: string
  checkFilesMap: checkFilesMap
  hasFilenames: boolean

  constructor (options: JsonBuilderOptions) {
    this.testSessionId = options.testSessionId
    this.numChecks = options.numChecks
    this.runLocation = options.runLocation
    this.checkFilesMap = options.checkFilesMap
    this.hasFilenames = !(options.checkFilesMap.size === 1 && options.checkFilesMap.has(undefined))
  }

  render (): string {
    const testSessionSummary: any = {
      testSessionId: this.testSessionId,
      numChecks: this.numChecks,
      runLocation: this.runLocation,
      checks: [],
    }

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    for (const [_, checkMap] of this.checkFilesMap.entries()) {
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      for (const [_, { result, testResultId, numRetries }] of checkMap.entries()) {
        const check: any = {
          result: result.hasFailures ? 'Fail' : 'Pass',
          name: result.name,
          checkType: result.checkType,
          durationMilliseconds: result.responseTime ?? null,
          filename: null,
          link: null,
          runError: result.runError || null,
          retries: numRetries,
        }

        if (this.hasFilenames) {
          check.filename = result.sourceFile
        }

        if (this.testSessionId && testResultId) {
          check.link = `${getTestSessionUrl(this.testSessionId)}/results/${testResultId}`
        }

        testSessionSummary.checks.push(check)
      }
    }

    return JSON.stringify(testSessionSummary, null, 2)
  }
}

export default class JsonReporter extends AbstractListReporter {
  onBegin (checks: Array<{ check: any, checkRunId: CheckRunId, sequenceId: SequenceId }>, testSessionId?: string) {
    super.onBegin(checks, testSessionId)
    printLn(`Running ${this.numChecks} checks in ${this._runLocationString()}.`, 2, 1)
  }

  onEnd () {
    this._printBriefSummary()
    const jsonBuilder = new JsonBuilder({
      testSessionId: this.testSessionId,
      numChecks: this.numChecks!,
      runLocation: this._runLocationString(),
      checkFilesMap: this.checkFilesMap!,
    })

    const json = jsonBuilder.render()

    const summaryFilename = process.env.CHECKLY_REPORTER_JSON_OUTPUT ?? outputFile
    fs.mkdirSync(path.resolve(path.dirname(summaryFilename)), { recursive: true })
    fs.writeFileSync(summaryFilename, json)

    printLn(`JSON report saved in '${path.resolve(summaryFilename)}'.`, 2)

    this._printTestSessionsUrl()
  }
}
