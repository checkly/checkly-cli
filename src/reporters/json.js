class JsonReporter {

}

module.exports = JsonReporter
