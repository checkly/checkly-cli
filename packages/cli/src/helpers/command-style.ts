import chalk from 'chalk'
import { ux } from '@oclif/core'

import { BaseCommand } from '../commands/baseCommand'
import { wrap } from './wrap'
import logSymbols from 'log-symbols'

const wrapOptions = {
  prefix: '  ',
  length: 78,
}

export class CommandStyle {
  private c: BaseCommand

  constructor (command: BaseCommand) {
    this.c = command
  }

  actionStart (message: string) {
    if (this.c.fancy) {
      ux.action.start(message, undefined, { stdout: true })
    }
  }

  actionSuccess () {
    if (this.c.fancy) {
      ux.action.stop(`✅`)
      this.c.log()
    }
  }

  actionFailure () {
    if (this.c.fancy) {
      ux.action.stop(`❌`)
      this.c.log()
    }
  }

  comment (message: string) {
    this.c.log(chalk.cyan(wrap(message, { ...wrapOptions, prefix: '// ' })))
    this.c.log()
  }

  longSuccess (title: string, message: string) {
    this.c.log(`${logSymbols.success} ${title}`)
    this.c.log()
    this.c.log(wrap(message, wrapOptions))
    this.c.log()
  }

  shortSuccess (message: string) {
    this.c.log(`${logSymbols.success} ${message}`)
    this.c.log()
  }

  longInfo (title: string, message: string) {
    this.c.log(`${logSymbols.info} ${title}`)
    this.c.log()
    this.c.log(wrap(message, wrapOptions))
    this.c.log()
  }

  shortInfo (message: string) {
    this.c.log(`${logSymbols.info} ${message}`)
    this.c.log()
  }

  longWarning (title: string, message: string) {
    this.c.log(`${logSymbols.warning} ${title}`)
    this.c.log()
    this.c.log(chalk.yellow(wrap(message, wrapOptions)))
    this.c.log()
  }

  shortWarning (message: string) {
    this.c.log(`${logSymbols.warning} ${chalk.yellow(message)}`)
    this.c.log()
  }

  longError (title: string, message: string) {
    this.c.log(`${logSymbols.error} ${title}`)
    this.c.log()
    this.c.log(chalk.red(wrap(message, wrapOptions)))
    this.c.log()
  }

  shortError (message: string) {
    this.c.log(`${logSymbols.error} ${chalk.red(message)}`)
    this.c.log()
  }

  fatal (message: string) {
    this.c.log(chalk.red(message))
    this.c.log()
  }
}
