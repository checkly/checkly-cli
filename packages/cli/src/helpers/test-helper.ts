import { PrivateRunLocation, RunLocation } from '../services/abstract-check-runner'
import { Session } from '../constructs'
import type { Region } from '..'
import { ReporterType } from '../reporters/reporter'
import { isCI } from 'ci-info'

const DEFAULT_REGION = 'eu-central-1'

export async function prepareRunLocation (
  configOptions: { runLocation?: keyof Region, privateRunLocation?: string } = {},
  cliFlags: { runLocation?: keyof Region, privateRunLocation?: string } = {},
  api: any,
  accountId: string,
): Promise<RunLocation> {
  // Command line options take precedence
  if (cliFlags.runLocation) {
    const { data: availableLocations } = await api.locations.getAll()
    if (availableLocations.some((l: { region: string | undefined }) => l.region === cliFlags.runLocation)) {
      return { type: 'PUBLIC', region: cliFlags.runLocation }
    }
    throw new Error(`Unable to run checks on unsupported location "${cliFlags.runLocation}". ` +
      `Supported locations are:\n${availableLocations.map((l: { region: any }) => `${l.region}`).join('\n')}`)
  } else if (cliFlags.privateRunLocation) {
    return preparePrivateRunLocation(cliFlags.privateRunLocation, api, accountId)
  } else if (configOptions.runLocation && configOptions.privateRunLocation) {
    throw new Error('Both runLocation and privateRunLocation fields were set in your Checkly config file.' +
      ` Please only specify one run location. The configured locations were' +
        ' "${configOptions.runLocation}" and "${configOptions.privateRunLocation}"`)
  } else if (configOptions.runLocation) {
    return { type: 'PUBLIC', region: configOptions.runLocation }
  } else if (configOptions.privateRunLocation) {
    return preparePrivateRunLocation(configOptions.privateRunLocation, api, accountId)
  } else {
    return { type: 'PUBLIC', region: DEFAULT_REGION }
  }
}

export async function preparePrivateRunLocation (privateLocationSlugName: string, api: any, accountId: string): Promise<PrivateRunLocation> {
  try {
    const privateLocations = await Session.getPrivateLocations()
    const privateLocation = privateLocations.find(({ slugName }) => slugName === privateLocationSlugName)
    if (privateLocation) {
      return { type: 'PRIVATE', id: privateLocation.id, slugName: privateLocationSlugName }
    }
    const { data: account } = await api.accounts.get(accountId)
    throw new Error(`The specified private location "${privateLocationSlugName}" was not found on account "${account.name}".`)
  } catch (err: any) {
    throw new Error(`Failed to get private locations. ${err.message}.`)
  }
}

export function prepareReportersTypes (reporterFlag: ReporterType, cliReporters: ReporterType[] = []): ReporterType[] {
  if (!reporterFlag && !cliReporters.length) {
    return [isCI ? 'ci' : 'list']
  }
  return reporterFlag ? [reporterFlag] : cliReporters
}
