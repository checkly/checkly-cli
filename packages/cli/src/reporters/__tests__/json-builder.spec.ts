import { JsonBuilder } from '../json'
import { generateMapAndTestResultIds } from './helpers'

const testSessionId = '0c4c64b3-79c5-44a6-ae07-b580ce73f328'
const runLocation = 'eu-west-1'
describe('JsonBuilder', () => {
  test('renders basic JSON output with no assets & links', () => {
    const checkFilesMap = generateMapAndTestResultIds({ includeTestResultIds: false })
    const json = new JsonBuilder({
      testSessionId: undefined,
      numChecks: checkFilesMap.size,
      runLocation,
      checkFilesMap,
    }).render()
    expect(json).toMatchSnapshot('json-basic')
  })
  test('renders basic JSON output with run errors', () => {
    const checkFilesMap = generateMapAndTestResultIds({ includeTestResultIds: false, includeRunErrors: true })
    const json = new JsonBuilder({
      testSessionId: undefined,
      numChecks: checkFilesMap.size,
      runLocation,
      checkFilesMap,
    }).render()
    expect(json).toMatchSnapshot('json-basic')
  })
  test('renders JSON markdown output with assets & links', () => {
    const checkFilesMap = generateMapAndTestResultIds({ includeTestResultIds: true })
    const json = new JsonBuilder({
      testSessionId,
      numChecks: checkFilesMap.size,
      runLocation,
      checkFilesMap,
    }).render()
    expect(json).toMatchSnapshot('json-with-assets-links')
  })
})
