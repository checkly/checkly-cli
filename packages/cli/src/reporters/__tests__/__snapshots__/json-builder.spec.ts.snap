// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsonBuilder renders JSON markdown output with assets & links: json-with-assets-links 1`] = `
"{
  "testSessionId": "0c4c64b3-79c5-44a6-ae07-b580ce73f328",
  "numChecks": 2,
  "runLocation": "eu-west-1",
  "checks": [
    {
      "result": "Pass",
      "name": "my-test.spec.ts",
      "checkType": "BROWSER",
      "durationMilliseconds": 6522,
      "filename": "src/__checks__/folder/browser.check.ts",
      "link": "https://app.checklyhq.com/test-sessions/0c4c64b3-79c5-44a6-ae07-b580ce73f328/results/702961fd-7e2c-45f0-97be-1aa9eabd4d82",
      "runError": "Run error",
      "retries": 0
    },
    {
      "result": "Pass",
      "name": "Test API check",
      "checkType": "API",
      "durationMilliseconds": 1234,
      "filename": "src/some-other-folder/api.check.ts",
      "link": "https://app.checklyhq.com/test-sessions/0c4c64b3-79c5-44a6-ae07-b580ce73f328/results/1c0be612-a5ec-432e-ac1c-837d2f70c010",
      "runError": "Run error",
      "retries": 0
    }
  ]
}"
`;

exports[`JsonBuilder renders basic JSON output with no assets & links: json-basic 1`] = `
"{
  "numChecks": 2,
  "runLocation": "eu-west-1",
  "checks": [
    {
      "result": "Pass",
      "name": "my-test.spec.ts",
      "checkType": "BROWSER",
      "durationMilliseconds": 6522,
      "filename": "src/__checks__/folder/browser.check.ts",
      "link": null,
      "runError": null,
      "retries": 0
    },
    {
      "result": "Pass",
      "name": "Test API check",
      "checkType": "API",
      "durationMilliseconds": 1234,
      "filename": "src/some-other-folder/api.check.ts",
      "link": null,
      "runError": null,
      "retries": 0
    }
  ]
}"
`;

exports[`JsonBuilder renders basic JSON output with run errors: json-basic 1`] = `
"{
  "numChecks": 2,
  "runLocation": "eu-west-1",
  "checks": [
    {
      "result": "Pass",
      "name": "my-test.spec.ts",
      "checkType": "BROWSER",
      "durationMilliseconds": 6522,
      "filename": "src/__checks__/folder/browser.check.ts",
      "link": null,
      "runError": "Run error",
      "retries": 0
    },
    {
      "result": "Pass",
      "name": "Test API check",
      "checkType": "API",
      "durationMilliseconds": 1234,
      "filename": "src/some-other-folder/api.check.ts",
      "link": null,
      "runError": "Run error",
      "retries": 0
    }
  ]
}"
`;
